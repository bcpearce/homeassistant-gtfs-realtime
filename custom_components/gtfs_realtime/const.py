"""Constants for the GTFS Realtime integration."""

DOMAIN = "gtfs_realtime"

CONF_GTFS_PROVIDER = "gtfs_provider"
CONF_API_KEY = "api_key"
CONF_GTFS_STATIC_DATA = "gtfs_static_data"
CONF_URL_ENDPOINTS = "url_endpoints"
CONF_ROUTE_ICONS = "route_icons"
CONF_ROUTE_IDS = "route_ids"
CONF_STOP_IDS = "stop_ids"
CONF_ARRIVAL_LIMIT = "arrival_limit"
STOP_ID = "stop_id"
ROUTE_ID = "route_id"

SSI_DB = "station_stop_info_db"
TI_DB = "trip_info_db"
CAL_DB = "calendar_db"
RTI_DB = "route_info_db"

COORDINATOR_REALTIME = "coordinator_realtime"
COORDINATOR_STATIC = "coordinator_static"

PLATFORM = "platform"

"""These Constants are user facing."""

TRIP_ID_PRETTY = "Trip ID"
HEADSIGN_PRETTY = "Headsign"
ROUTE_COLOR_PRETTY = "Route Color"
ROUTE_TEXT_COLOR_PRETTY = "Route Text Color"
ROUTE_TYPE_PRETTY = "Route Type"

HEADER_PRETTY = "Header"
DESCRIPTION_PRETTY = "Description"
