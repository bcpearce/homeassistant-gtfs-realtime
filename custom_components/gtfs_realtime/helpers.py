"""Helpers for GTFS Realtime."""


def header_dict_from_header_str(header: str) -> dict[str, str] | None:
    if header is None:
        return None
    return dict([[x.strip() for x in header.split(":")]])
