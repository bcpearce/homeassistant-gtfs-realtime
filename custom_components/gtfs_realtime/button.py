"""Button support for GTFS Schedule Update."""

from __future__ import annotations

from collections.abc import Callable, Coroutine
from dataclasses import dataclass
from functools import cached_property
import os

from homeassistant.components.button import (
    ButtonDeviceClass,
    ButtonEntity,
    ButtonEntityDescription,
)
from homeassistant.const import EntityCategory
from homeassistant.core import HomeAssistant
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from custom_components.gtfs_realtime import GtfsRealtimeConfigEntry
from custom_components.gtfs_realtime.const import DOMAIN

from .coordinator import GtfsRealtimeCoordinator


@dataclass(frozen=True, kw_only=True)
class GtfsRealtimeButtonDescription(ButtonEntityDescription):
    """GTFS Realtime Button description."""

    press_action: Callable[[GtfsRealtimeCoordinator, os.PathLike], Coroutine]


async def handle_refresh_button_press(
    coordinator: GtfsRealtimeCoordinator, gtfs_static_source: os.PathLike
):
    """Called on pressing refresh feed entity."""
    coordinator.static_update_targets.add(gtfs_static_source)
    await coordinator.async_update_static_data()


BUTTONS: dict[str, GtfsRealtimeButtonDescription] = {
    "refresh": GtfsRealtimeButtonDescription(
        key="refresh",
        translation_key="refresh",
        device_class=ButtonDeviceClass.UPDATE,
        entity_category=EntityCategory.CONFIG,
        press_action=handle_refresh_button_press,
    ),
    "clear": GtfsRealtimeButtonDescription(
        key="clear",
        translation_key="clear",
        device_class=ButtonDeviceClass.UPDATE,
        entity_category=EntityCategory.CONFIG,
        press_action=lambda coordinator: coordinator.async_update_static_data(
            clear_old_data=True
        ),
    ),
}


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: GtfsRealtimeConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    coordinator = config_entry.runtime_data
    async_add_entities(
        [
            ScheduleUpdateButton(coordinator, BUTTONS["refresh"], gtfs_static_source)
            for gtfs_static_source in coordinator.gtfs_static_zip
        ]
        + [ScheduleClearButton(coordinator, BUTTONS["clear"])]
    )


class ScheduleButtonBase(ButtonEntity, CoordinatorEntity):
    entity_description: GtfsRealtimeButtonDescription

    def __init__(
        self,
        coordinator: GtfsRealtimeCoordinator,
        description: GtfsRealtimeButtonDescription,
    ):
        super().__init__(coordinator)
        self.entity_description = description

    @cached_property
    def device_info(self) -> DeviceInfo:
        """Return the device info."""
        return DeviceInfo(
            identifiers={(DOMAIN, ",".join(self.coordinator.gtfs_static_zip))},
            name="GTFS Schedule",
            manufacturer=self.coordinator.gtfs_provider,
        )


class ScheduleUpdateButton(ScheduleButtonBase):
    def __init__(
        self,
        coordinator: GtfsRealtimeCoordinator,
        description: GtfsRealtimeButtonDescription,
        gtfs_static_source: os.PathLike | None = None,
    ):
        super().__init__(coordinator, description)
        self.gtfs_static_source = gtfs_static_source
        self._attr_unique_id = f"refresh_gtfs_schedule_button-{gtfs_static_source}"
        self._attr_translation_key = description.translation_key
        # Figure out how to use translations for this
        self._attr_name = f"Refresh Schedule Feed: {self.gtfs_static_source}"

    @cached_property
    def extra_state_attributes(self) -> dict[str, str]:
        return {"target_url": str(self.gtfs_static_source)}

    async def async_press(self) -> None:
        """Trigger the button action."""
        await self.entity_description.press_action(
            self.coordinator, self.gtfs_static_source
        )


class ScheduleClearButton(ScheduleButtonBase):
    entity_description: GtfsRealtimeButtonDescription

    def __init__(
        self,
        coordinator: GtfsRealtimeCoordinator,
        description: GtfsRealtimeButtonDescription,
    ):
        super().__init__(coordinator, description)
        self._attr_unique_id = (
            f"clear_gtfs_schedule-{"-".join(coordinator.gtfs_static_zip)}"
        )
        self._attr_translation_key = description.translation_key
        self._attr_name = "Clear GTFS Schedule"

    @cached_property
    def extra_state_attributes(self) -> dict[str, str]:
        return {
            f"target_url {i+1}": url
            for i, url in enumerate(self.coordinator.gtfs_static_zip)
        }

    async def async_press(self) -> None:
        """Trigger the button action."""
        await self.entity_description.press_action(self.coordinator)
