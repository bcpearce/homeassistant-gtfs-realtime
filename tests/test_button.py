"""Test button."""

from unittest.mock import AsyncMock, patch

from homeassistant.components.button import (
    DOMAIN as BUTTON_DOMAIN,
    SERVICE_PRESS as BUTTON_SERVICE_PRESS,
)
from homeassistant.const import ATTR_ENTITY_ID
from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_registry as er
import pytest
from pytest_homeassistant_custom_component.common import MockConfigEntry


@pytest.fixture
def async_update_patcher():
    return patch(
        "custom_components.gtfs_realtime.coordinator.GtfsRealtimeCoordinator.async_update_static_data",  # noqa E501
        new_callable=AsyncMock,
    )


async def test_setup_button(
    hass: HomeAssistant, entry_v2_full: MockConfigEntry, async_update_patcher, snapshot
):
    """Test setting ups buttons in integration."""
    with (
        patch(
            "custom_components.gtfs_realtime.coordinator.GtfsRealtimeCoordinator._async_update_data",  # noqa E501
            new_callable=AsyncMock,
        ),
        async_update_patcher,
    ):
        entry_v2_full.add_to_hass(hass)
        assert await hass.config_entries.async_setup(entry_v2_full.entry_id)
        await hass.async_block_till_done()

        ent_reg = er.async_get(hass)
        button_ids = [k for k, v in ent_reg.entities.items() if k.startswith("button")]
        assert button_ids == snapshot


# @pytest.mark.skip("Not working as expected")
async def test_button_press(
    hass: HomeAssistant, entry_v2_full: MockConfigEntry, async_update_patcher
):
    """Test refreshing static feed."""
    with (
        patch(
            "custom_components.gtfs_realtime.coordinator.GtfsRealtimeCoordinator._async_update_data",  # noqa E501
            new_callable=AsyncMock,
        ),
        async_update_patcher,
    ):
        entry_v2_full.add_to_hass(hass)
        assert await hass.config_entries.async_setup(entry_v2_full.entry_id)

    with async_update_patcher as update_static_data_mock:
        await hass.services.async_call(
            BUTTON_DOMAIN,
            BUTTON_SERVICE_PRESS,
            {
                ATTR_ENTITY_ID: "button.refresh_schedule_feed_https_example_com_gtfs1_zip"
            },
            blocking=True,
        )
        update_static_data_mock.assert_called()
        await hass.services.async_call(
            BUTTON_DOMAIN,
            BUTTON_SERVICE_PRESS,
            {ATTR_ENTITY_ID: "button.clear_gtfs_schedule"},
            blocking=True,
        )
        update_static_data_mock.assert_called_with(clear_old_data=True)
