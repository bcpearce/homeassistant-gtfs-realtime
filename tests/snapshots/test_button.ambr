# serializer version: 1
# name: test_setup_button
  list([
    'button.refresh_schedule_feed_https_example_com_gtfs1_zip',
    'button.refresh_schedule_feed_https_example_com_gtfs2_zip',
    'button.clear_gtfs_schedule',
  ])
# ---
